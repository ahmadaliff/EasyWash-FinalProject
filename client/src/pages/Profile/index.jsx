import PropTypes from 'prop-types';
import { useEffect, useRef, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import config from '@config/index';

import EditProfile from '@components/EditProfile';

import { Edit, FlipCameraIos } from '@mui/icons-material';
import { Avatar, Card, CardContent, Skeleton } from '@mui/material';

import { actionEditPhotoProfile, actionGetProfile, actionResetProfile } from '@pages/Profile/actions';
import { selectProfile } from '@pages/Profile/selectors';
import { selectUser } from '@containers/Client/selectors';

import classes from '@pages/Profile/style.module.scss';

const Profile = ({ user, profile }) => {
  const dispatch = useDispatch();
  const fileInput = useRef(null);
  const [loading, setLoading] = useState(true);
  const [isDialogOpen, setIsDialogOpen] = useState(false);

  useEffect(() => {
    if (!profile) dispatch(actionGetProfile(user?.id));
    return () => {
      if (profile) {
        dispatch(actionResetProfile());
      }
    };
  }, [dispatch, profile, user?.id]);

  const handleImageChange = (e) => {
    dispatch(actionEditPhotoProfile(e.target.files[0]));
  };

  return (
    <main className={classes.mainWrap} data-testid="profile-page">
      <div className={classes.Wrap}>
        <input
          type="file"
          ref={fileInput}
          style={{ display: 'none' }}
          onChange={handleImageChange}
          accept="image/*"
          data-testid="input-image"
        />
        <Card>
          <CardContent className={classes.cardWrap}>
            <div
              className={classes.imgWrap}
              onClick={() => fileInput.current.click()}
              data-testid="button-change-image"
            >
              {profile?.imagePath ? (
                <>
                  {loading && <Skeleton variant="square" className={classes.skeleton} />}
                  <img
                    src={
                      profile?.imagePath.includes('https://lh3.googleusercontent.com/')
                        ? profile?.imagePath
                        : `${config.api.server}${profile?.imagePath}`
                    }
                    alt={profile?.fullName}
                    loading="lazy"
                    onLoad={() => setLoading(false)}
                  />
                </>
              ) : (
                <Avatar className={classes.avatar} variant="square">
                  {profile?.fullName?.split(' ')[0][0]}
                  {profile?.fullName?.split(' ').length > 1 && profile?.fullName?.split(' ')[1][0]}
                </Avatar>
              )}
              <Edit className={classes.buttonEdit} />
              <FlipCameraIos className={classes.iconCamera} />
            </div>
            <button
              type="button"
              onClick={() => setIsDialogOpen(true)}
              className={classes.buttonEditProfile}
              data-testid="button-edit"
            >
              <FormattedMessage id="app_profile_edit" />
            </button>
          </CardContent>
        </Card>
        <div>
          <h3>
            <FormattedMessage id="app_profile" />
          </h3>
          <span className={classes.profileInfo}>
            <p className={classes.profileLabel}>
              <FormattedMessage id="app_user_fullName" />
            </p>
            <p className={classes.profileText}>{profile?.fullName}</p>
          </span>
          <span className={classes.profileInfo}>
            <p className={classes.profileLabel}>
              <FormattedMessage id="app_user_phone" />
            </p>
            <p className={classes.profileText}>{profile?.phone || '-'}</p>
          </span>
          <span className={classes.profileInfo}>
            <p className={classes.profileLabel}>
              <FormattedMessage id="app_user_email" />
            </p>
            <p className={classes.profileText}>{profile?.email}</p>
          </span>
          <span className={classes.profileInfo}>
            <p className={classes.profileLabel}>
              <FormattedMessage id="app_account_type" />
            </p>
            <p className={classes.profileText}>{profile?.role}</p>
          </span>
        </div>
      </div>
      <EditProfile profile={profile} open={isDialogOpen} handleClose={() => setIsDialogOpen(false)} />
    </main>
  );
};

Profile.propTypes = {
  user: PropTypes.object,
  profile: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  user: selectUser,
  profile: selectProfile,
});

export default connect(mapStateToProps)(Profile);
