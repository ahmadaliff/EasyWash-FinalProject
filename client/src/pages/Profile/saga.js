import { takeLatest, call, put } from 'redux-saga/effects';
import toast from 'react-hot-toast';
import CryptoJS from 'crypto-js';

import { apiHandleEditPhotoProfile, apiHandleEditProfile, apiHandleGetProfile } from '@domain/api';

import intlHelper from '@utils/intlHelper';

import config from '@config/index';

import { showPopup, setLoading } from '@containers/App/actions';
import { EDIT_PHOTO_PROFILE, EDIT_PROFILE, GET_PROFILE } from '@pages/Profile/constants';
import { actionSetProfile } from '@pages/Profile/actions';

function* sagaHandleGetUser() {
  yield put(setLoading(true));
  try {
    const response = yield call(apiHandleGetProfile);
    yield put(actionSetProfile(response.data));
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404 || error?.response?.status === 401) {
      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

function* sagaHandleEditPhotoProfile({ data }) {
  yield put(setLoading(true));
  try {
    const response = yield call(apiHandleEditPhotoProfile, { image: data });
    yield put(actionSetProfile(response.data));
    toast.success(intlHelper({ message: response?.message }));
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404 || error?.response?.status === 401) {
      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

function* sagaHandleEditProfile({ data }) {
  yield put(setLoading(true));
  try {
    if (data?.new_password || data?.old_password) {
      data.new_password = CryptoJS.AES.encrypt(data.new_password, config.api.secretKeyCrypto).toString();
      data.old_password = CryptoJS.AES.encrypt(data.old_password, config.api.secretKeyCrypto).toString();
    }
    const response = yield call(apiHandleEditProfile, data);
    toast.success(intlHelper({ message: response?.message }));
    yield put(actionSetProfile(response.data));
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404 || error?.response?.status === 401) {
      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

export default function* profileSaga() {
  yield takeLatest(GET_PROFILE, sagaHandleGetUser);
  yield takeLatest(EDIT_PHOTO_PROFILE, sagaHandleEditPhotoProfile);
  yield takeLatest(EDIT_PROFILE, sagaHandleEditProfile);
}
