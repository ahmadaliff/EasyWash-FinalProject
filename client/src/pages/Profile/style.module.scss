@import '@styles/base';

.mainWrap {
  --image-size-sm: 10rem;
  --image-size-md: 12rem;
  @include directionY;
  padding: 5rem;
  min-height: calc(100vh - 3.55rem);
  align-items: center;
  box-sizing: border-box;
  position: relative;
  * {
    box-sizing: border-box;
  }
  .Wrap {
    @include directionY;
    @include apply-at-md-size {
      @include directionX;
    }
    gap: 0.75rem;
    h2 {
      margin: auto;
      font-size: 24px;
      font-weight: 700;
      text-align: center;
    }
    .cardWrap {
      @include directionY;
      gap: 0.5rem;
      .imgWrap {
        height: var(--image-size-sm);
        width: var(--image-size-sm);
        @include apply-at-md-size {
          height: var(--image-size-md);
          width: var(--image-size-md);
        }
        overflow: hidden;
        margin: auto;
        position: relative;
        img {
          object-fit: cover;
          object-position: top;
          height: var(--image-size-sm);
          width: var(--image-size-sm);
          @include apply-at-md-size {
            height: var(--image-size-md);
            width: var(--image-size-md);
          }
        }
        .skeleton {
          position: absolute;
          height: var(--image-size-sm);
          width: var(--image-size-sm);
          @include apply-at-md-size {
            height: var(--image-size-md);
            width: var(--image-size-md);
          }
        }
        .avatar {
          height: var(--image-size-sm);
          width: var(--image-size-sm);
          @include apply-at-md-size {
            height: var(--image-size-md);
            width: var(--image-size-md);
          }
        }
        .buttonEdit {
          position: absolute;
          display: none;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
        }
        &:hover {
          cursor: pointer;
          transform: scale(1.025, 1.025);
          opacity: 0.7;
        }
        &:hover > .buttonEdit {
          display: block;
        }
      }
      .buttonEditProfile {
        width: 100%;
        background-color: var(--color-bg);
        color: var(--color-primary);
        border: 1px solid var(--color-primary);
        border-radius: 2.5rem;
        padding: 0.5rem 2.5rem 0.5rem 2.5rem;
        margin: auto;
        margin-top: 0.75rem;
        box-shadow: none;
        font-weight: bold;
        cursor: pointer;
        &:hover {
          transform: scale(1.025, 1.025);
        }
      }
    }
  }
  .profileInfo {
    @include directionX;
    align-items: center;
    font-size: 0.75rem;
    font-weight: 500;
    .profileLabel {
      float: left;
      width: 10rem;
    }
    .profileText {
      width: 100%;
    }
  }
}
