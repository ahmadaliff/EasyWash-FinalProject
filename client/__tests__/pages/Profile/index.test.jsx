import Profile from '@pages/Profile';
import { render, fireEvent, act } from '@utils/testHelper';

let wrapper;
const mockNavigate = jest.fn();
const mockDispatch = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));
jest.mock('@pages/Profile/selectors', () => ({
  selectProfile: () => ({
    imagePath: null,
    fullName: 'ahmad alif',
    phone: '03313131831',
    email: 'email@email.com',
    role: 'admin',
  }),
}));
jest.mock('@components/MapLeaflet', () => jest.fn().mockImplementation(() => <div />));

beforeEach(() => {
  act(() => {
    wrapper = render(<Profile />);
  });
});

describe('Profile Page', () => {
  test('render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('profile-page')).toBeInTheDocument();
  });
  test('navigate back when back button click', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-change-image');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(getByTestId('input-image')).toBeInTheDocument();
  });
  test('should be open dialog edit', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-edit');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(getByTestId('dialog-edit-profile')).toBeInTheDocument();
  });
  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
