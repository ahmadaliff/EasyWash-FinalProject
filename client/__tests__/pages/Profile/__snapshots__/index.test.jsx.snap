// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Page Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style=""
  >
    <div>
      <main
        class="mainWrap"
        data-testid="profile-page"
      >
        <div
          class="Wrap"
        >
          <input
            accept="image/*"
            data-testid="input-image"
            style="display: none;"
            type="file"
          />
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardContent-root cardWrap css-46bh2p-MuiCardContent-root"
            >
              <div
                class="imgWrap"
                data-testid="button-change-image"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-square MuiAvatar-colorDefault avatar css-18gxrea-MuiAvatar-root"
                >
                  a
                  a
                </div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium buttonEdit css-i4bv87-MuiSvgIcon-root"
                  data-testid="EditIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium iconCamera css-i4bv87-MuiSvgIcon-root"
                  data-testid="FlipCameraIosIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 5h-3.17L15 3H9L7.17 5H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm-8 13c-2.76 0-5-2.24-5-5H5l2.5-2.5L10 13H8c0 2.21 1.79 4 4 4 .58 0 1.13-.13 1.62-.35l.74.74c-.71.37-1.5.61-2.36.61zm4.5-2.5L14 13h2c0-2.21-1.79-4-4-4-.58 0-1.13.13-1.62.35l-.74-.73C10.35 8.24 11.14 8 12 8c2.76 0 5 2.24 5 5h2l-2.5 2.5z"
                  />
                </svg>
              </div>
              <button
                class="buttonEditProfile"
                data-testid="button-edit"
                type="button"
              >
                Ubah Profil
              </button>
            </div>
          </div>
          <div>
            <h3>
              Profil
            </h3>
            <span
              class="profileInfo"
            >
              <p
                class="profileLabel"
              >
                Nama Lengkap
              </p>
              <p
                class="profileText"
              >
                ahmad alif
              </p>
            </span>
            <span
              class="profileInfo"
            >
              <p
                class="profileLabel"
              >
                Telp/No.Hp
              </p>
              <p
                class="profileText"
              >
                03313131831
              </p>
            </span>
            <span
              class="profileInfo"
            >
              <p
                class="profileLabel"
              >
                Email
              </p>
              <p
                class="profileText"
              >
                email@email.com
              </p>
            </span>
            <span
              class="profileInfo"
            >
              <p
                class="profileLabel"
              >
                Tipe Akun
              </p>
              <p
                class="profileText"
              >
                admin
              </p>
            </span>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="mainWrap"
      data-testid="profile-page"
    >
      <div
        class="Wrap"
      >
        <input
          accept="image/*"
          data-testid="input-image"
          style="display: none;"
          type="file"
        />
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root cardWrap css-46bh2p-MuiCardContent-root"
          >
            <div
              class="imgWrap"
              data-testid="button-change-image"
            >
              <div
                class="MuiAvatar-root MuiAvatar-square MuiAvatar-colorDefault avatar css-18gxrea-MuiAvatar-root"
              >
                a
                a
              </div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium buttonEdit css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium iconCamera css-i4bv87-MuiSvgIcon-root"
                data-testid="FlipCameraIosIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 5h-3.17L15 3H9L7.17 5H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm-8 13c-2.76 0-5-2.24-5-5H5l2.5-2.5L10 13H8c0 2.21 1.79 4 4 4 .58 0 1.13-.13 1.62-.35l.74.74c-.71.37-1.5.61-2.36.61zm4.5-2.5L14 13h2c0-2.21-1.79-4-4-4-.58 0-1.13.13-1.62.35l-.74-.73C10.35 8.24 11.14 8 12 8c2.76 0 5 2.24 5 5h2l-2.5 2.5z"
                />
              </svg>
            </div>
            <button
              class="buttonEditProfile"
              data-testid="button-edit"
              type="button"
            >
              Ubah Profil
            </button>
          </div>
        </div>
        <div>
          <h3>
            Profil
          </h3>
          <span
            class="profileInfo"
          >
            <p
              class="profileLabel"
            >
              Nama Lengkap
            </p>
            <p
              class="profileText"
            >
              ahmad alif
            </p>
          </span>
          <span
            class="profileInfo"
          >
            <p
              class="profileLabel"
            >
              Telp/No.Hp
            </p>
            <p
              class="profileText"
            >
              03313131831
            </p>
          </span>
          <span
            class="profileInfo"
          >
            <p
              class="profileLabel"
            >
              Email
            </p>
            <p
              class="profileText"
            >
              email@email.com
            </p>
          </span>
          <span
            class="profileInfo"
          >
            <p
              class="profileLabel"
            >
              Tipe Akun
            </p>
            <p
              class="profileText"
            >
              admin
            </p>
          </span>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
